tests/js/lib/link-checker.js

```diff
@@ -0,0 +1,166 @@
+
+        var report = fluid.model.transformWithRules(that, that.options.rules.report);
+        that.events.onResultsAvailable.fire(report);
+    }
+};
+
+fluid.tests.docs.linkChecker.urlBeforeHash = function (link) {
+    return link.indexOf("#") !== -1 ? link.substring(0, link.indexOf("#")) : link;
+};
+
+fluid.tests.docs.linkChecker.urlAfterHash = function (link) {
+    return link.indexOf("#") !== -1 ? link.substring(link.indexOf("#") + 1) : "";
+};
+
+fluid.tests.docs.linkChecker.checkInternalLinks = function (that) {
+    fluid.each(that.internalLinksToCheck, function (link) {
+        var linkBeforeHash = fluid.tests.docs.linkChecker.urlBeforeHash(link);
+        var linkAfterHash  = fluid.tests.docs.linkChecker.urlAfterHash(link);
+        if (linkAfterHash && linkAfterHash.length) {
+            var pageIds = that.idsByPage[linkBeforeHash];
+            if (!pageIds || pageIds.indexOf(linkAfterHash) === -1) {
+                that.brokenInternalLinks.push(link);
+            }
+        }
+    });
+};
+
+fluid.tests.docs.linkChecker.checkPageLinks = function (that) {
+    fluid.each(that.scannedPages, function (results, page) {
+        if (results.error || results.statusCode !== 200) {
+            that.brokenPageLinks.push(page);
+        }
+    });
+};
+
+fluid.tests.docs.linkChecker.resolveSafely = function (baseUrl, path) {
+    var cleanPath = path.replace(/^about:blank/i, "");
+    return url.resolve(baseUrl, cleanPath);
+};
+
+fluid.tests.docs.linkChecker.processSingleResponse = function (that, pageToScan, error, response, body) {
+    var pageResults = { statusCode: response.statusCode};
+    if (error) {
+        pageResults.error = error;
+    }
+    else if (body) {
+        var dom = new jsdom.JSDOM(body);
+
+        // Add the page's elements with IDs to our map of known placeholders.
+        var pageIdNodeList = dom.window.document.querySelectorAll("[id]");
+        var pageIds = [];
+        for (var b = 0; b < pageIdNodeList.length; b++) {
+            pageIds.push(pageIdNodeList.item(b).id);
+        }
+        that.idsByPage[pageToScan] = pageIds;
+
+        var linksNodeList = dom.window.document.querySelectorAll("a[href]");
+        for (var a = 0; a < linksNodeList.length; a++) {
+            var link = linksNodeList.item(a);
+            var href = link.href;
+
+            if (href && href.length) {
+                var resolvedHref = fluid.tests.docs.linkChecker.resolveSafely(pageToScan, href);
+
+                var linkBeforeHash = fluid.tests.docs.linkChecker.urlBeforeHash(resolvedHref);
+                var linkAfterHash  = fluid.tests.docs.linkChecker.urlAfterHash(resolvedHref);
+
+                // As long as this is a placeholder link somewhere within the site, hold on to it to check later.
+                if (linkAfterHash.length && resolvedHref.indexOf(that.options.baseUrl) === 0) {
+                    that.internalLinksToCheck.push(resolvedHref);
```

Comment:
This is the point at which it would be helpful to push a little structure including pageToScan as well as just the URL

Reply:
I rethought my approach and realised that we don't care whether the link is internal or external, only whether it is broken and where it appears.  I addressed this in a slightly different way than you suggest, but the link checker now outputs details about which files fail.  To demonstrate, I added a broken link to a single file, which resulted in output like:

```
11:34:55.917:  jq: FAIL: Module "Checking links within docpad-generated site..." Test name "Checking links..." - Message: There should be no broken links... - at sequence position 4 of 4
11:34:55.917:  jq: Expected: 0
11:34:55.917:  jq: Actual: 1
11:34:55.917:  jq: Source:     at Object.assertEquals (/Users/duhrer/Source/rtf/infusion-docs/node_modules/infusion/tests/test-core/jqUnit/js/jqUnit.js:137:19)
    at fluid.test.docs.docpadLinks.checkResults (/Users/duhrer/Source/rtf/infusion-docs/tests/js/test-docpad-links.js:27:12)
    at togo (/Users/duhrer/Source/rtf/infusion-docs/node_modules/infusion/tests/test-core/utils/js/IoCTestUtils.js:202:33)
    at wrapped (/Users/duhrer/Source/rtf/infusion-docs/node_modules/infusion/tests/test-core/utils/js/IoCTestUtils.js:320:26)
    at Object.fire (/Users/duhrer/Source/rtf/infusion-docs/node_modules/infusion/src/framework/core/js/Fluid.js:1534:40)
    at fluid.tests.docs.linkChecker.scanSinglePageOrFinish (/Users/duhrer/Source/rtf/infusion-docs/tests/js/lib/link-checker.js:59:40)
    at Array.invokeInvoker (/Users/duhrer/Source/rtf/infusion-docs/node_modules/infusion/src/framework/core/js/FluidIoC.js:1741:29)
    at Object.that.complete (/Users/duhrer/Source/rtf/infusion-docs/node_modules/infusion/src/framework/core/js/FluidProm .... [output suppressed at 1024 chars - for more output, increase fluid.logObjectRenderChars]
11:34:55.918:  File 'http://localhost:7777/infusion/development/SwitchAPI.html' has a broken link to 'http://localhost:7777/infusion/development/SwitchAPI.html#this-doesnt-exist'.
```